source/server/listener_impl.h

```diff
@@ -204,6 +209,69 @@ class PerListenerFactoryContextImpl : public Configuration::ListenerFactoryConte
   ListenerImpl& listener_impl_;
 };
 
+/**
+ * PerFilterChainRebuilder is used to rebuild filter chain placeholder. It assigns an init manager
+ * to track filter chain dependencies and acts as the filter chain factory context creator.
+ * When rebuilding is complete or reaches timeout, it will send callback to workers to retry
+ * connections.
+ */
+class PerFilterChainRebuilder : public FilterChainFactoryContextCreator,
+                                Logger::Loggable<Logger::Id::config> {
+public:
+  PerFilterChainRebuilder(ListenerImpl& listener,
+                          const envoy::config::listener::v3::FilterChain* const& filter_chain,
+                          Configuration::FactoryContext& factory_context);
+
+  // FilterChainFactoryContextCreator
+  Configuration::FilterChainFactoryContextPtr createFilterChainFactoryContext(
+      const ::envoy::config::listener::v3::FilterChain* const filter_chain) override;
+
+  void storeWorkerInCallbackList(const std::string& worker_name);
+  void callbackToWorkers(bool success);
+  bool rebuildingFailed() { return state_ == State::Failed; }
+  bool rebuildUnfinished() { return state_ == State::Running; }
+  bool timeoutEnabled() { return timeout_enabled_; }
+  Init::Manager& initManager() { return *rebuild_init_manager_; }
+  void startRebuilding();
+  void stopRebuilding();
+
+private:
+  void startTimer();
+  void onTimeout();
+
+  ListenerImpl& listener_;
+  const envoy::config::listener::v3::FilterChain* const& filter_chain_;
+  Configuration::FactoryContext& parent_context_;
+
+  enum class State {
+    /**
+     * Rebuilding is still in progress.
+     */
+    Running,
+    /**
+     * Rebuilding has been completed successfully.
+     */
+    Succeeded,
+    /**
+     * Rebuilding has failed due to timeout or error.
+     */
+    Failed
+  };
+  State state_;
+
+  std::unique_ptr<Init::Manager> rebuild_init_manager_;
+  Init::WatcherImpl rebuild_watcher_;
+
+  bool timeout_enabled_{false};
```

Comment:
is timeout_enabled_ duplicating rebuild_timeout_?

Reply:
This flag is meant to help check whether to ```disableTimer``` when watcher is ready. It is kind of strange if we disable the timer when it is even not started.